[AMDGPU] Exclude certain opcodes from being marked as single use

ScottEgerton opened the pull request:
The s_singleuse_vdst instruction is used to mark regions of instructions that
produce values that have only one use.
Certain instructions take more than one cycle to execute, resulting in regions
being incorrectly marked.
This patch excludes these multi-cycle instructions from being marked as either
producing single use values or consuming single use values or both depending
on the instruction.

ScottEgerton committed:
[AMDGPU] Exclude certain opcodes from being marked as single use

The s_singleuse_vdst instruction is used to mark regions of instructions that
produce values that have only one use.
Certain instructions take more than one cycle to execute, resulting in regions
being incorrectly marked.
This patch excludes these multi-cycle instructions from being marked as either
producing single use values or consuming single use values or both depending
on the instruction.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:
```diff
@@ -132,6 +132,153 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
 
   AMDGPUInsertSingleUseVDST() : MachineFunctionPass(ID) {}
 
+  static bool IsValidOpcode(const MachineInstr &MI) {
```
This feels like it needs to be an instruction bit and/or searchable table

arsenm replied:
Or could we derive this from the scheduling info?

jayfoad replied:
Searchable table sounds good. We can't derive it from the sched model because it's not really related to how many cycles the instruction takes to execute; we should treat it as a separate list of instructions defined by the hardware design.

ScottEgerton replied:
I've changed this to be a searchable table.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:
```diff
@@ -132,6 +132,153 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
+    }
+  }
+
+  static bool IsValidConsumerOpcode(const MachineInstr &MI) {
```
Start with lowercase

ScottEgerton replied:
Done

ScottEgerton committed:
Convert single use exception definitions to TableGen

Previously invalid single use consumers and invalid single use
producers were listed in a C++ switch statement.
This commit changes this instead be defined in TableGen
alongside the instruction definitions
and accessed via a searchable table.

jayfoad commented on llvm/lib/Target/AMDGPU/VOPInstructions.td:
```diff
@@ -819,9 +825,7 @@ class VOP3P_DPPe_Common<bits<7> op, VOPProfile P> : VOP3P_DPPe_Common_Base<op, P
-  InstSI <P.OutsDPP, Ins, OpName#asmOps, pattern>,
-  VOP <OpName>,
-  SIMCInstr <OpName#"_dpp", SIEncodingFamily.NONE> {
+  VOP_Pseudo<OpName, "_dpp", P, P.OutsDPP, Ins, asmOps, pattern> {
```
Why this change - is it just a cleanup?

ScottEgerton replied:
This change allows dpp instructions to be included in the searchable table by inheriting the IsInvalidSingleUse* bits from VOP_Pseudo.

jayfoad replied:
I see. I think this just allows you to use VOP_Pseudo as a common superclass for all VOP instructions, DPP or not, when you define the searchable table. Perhaps you could use an even more common class like InstSI - I'm not sure if that would make the generated table bigger.

@Sisyph @kosarev any thoughts on this part of the patch?

kosarev replied:
Looks reasonable to me and a refinement as well. I second Jay's words on whether it makes sense to limit the flags to `VOP_Pseudo`s and then have explicit `SIInstrInfo::isVAL()` checks, though.

jayfoad replied:
As a followup I guess we could remove a lot of the definitions below like `let isPseudo = 1` that can now be inherited from `VOP_Pseudo`.

jayfoad commented on llvm/lib/Target/AMDGPU/VOP3Instructions.td:
```diff
@@ -156,10 +156,12 @@ defm V_MAX_F64 : VOP3Inst <"v_max_f64", VOP3_Profile<VOP_F64_F64_F64>, fmaxnum_l
 } // End SchedRW = [WriteDoubleAdd]
 
 let SchedRW = [WriteIntMul] in {
+defm V_MUL_LO_I32 : VOP3Inst <"v_mul_lo_i32", V_MUL_PROF<VOP_I32_I32_I32>>;
```
You don't need to handle V_MUL_LO_I32 differently from the other mul opcodes here. It was never generated because its behaviour was always identical to the U32 version, so recent generations have removed support for it, so it does not matter if you set the IsInvalid flag on it.

ScottEgerton replied:
That's good to know. I've changed the code back to handle all the mul opcodes together.

jayfoad commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:
```diff
@@ -607,6 +615,16 @@ bool isTrue16Inst(unsigned Opc) {
 
+bool isInvalidSingleUseConsumerInst(unsigned Opc) {
+  const SingleUseExceptionInfo *Info = getSingleUseExceptionHelper(Opc);
+  return Info ? Info->IsInvalidSingleUseConsumer : false;
```
Nit: `return Info && Info->IsInvalidSingleUseConsumer`. (I realize you were just copying the existing code.)

ScottEgerton replied:
Done

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:
```diff
@@ -132,6 +133,14 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
 
   AMDGPUInsertSingleUseVDST() : MachineFunctionPass(ID) {}
 
+  static bool isValidProducerInst(const MachineInstr &MI) {
```
Probably simpler to inline this function into its only call site. Then we would not have a mix of "isValid" and "isInvalid" functions, which I find a little confusing.

ScottEgerton replied:
Done

jayfoad commented on llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:
```diff
@@ -1,6 +1,5 @@
 # NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
 # RUN: llc -mtriple=amdgcn -mcpu=gfx1150 -mattr=+wavefrontsize32,-wavefrontsize64 -verify-machineinstrs -run-pass=amdgpu-insert-single-use-vdst %s -o - | FileCheck %s
```
If you don't care about the wave size in this test, remove the -mattr option.

ScottEgerton replied:
Done

jayfoad reviewed (commented):
I think this looks good overall, thanks.

ScottEgerton committed:
fixup! Convert single use exception definitions to TableGen

jayfoad reviewed (approved):
LGTM, thanks!

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.td:
```diff
@@ -2266,6 +2266,8 @@ class VOPProfile <list<ValueType> _ArgVT, bit _EnableClamp = 0> {
   field bit EnableClamp = _EnableClamp;
   field bit IsTrue16 = 0;
   field bit IsRealTrue16 = 0;
+  field bit IsInvalidSingleUseConsumer = 0;
+  field bit IsInvalidSingleUseProducer = 0;
```
Looking again, there is no reason for these fields to exist in VOPProfile.

jayfoad commented on llvm/lib/Target/AMDGPU/VOPInstructions.td:
```diff
@@ -175,6 +179,8 @@ class VOP3P_Pseudo <string opName, VOPProfile P, list<dag> pattern = []> :
 class VOP_Real<VOP_Pseudo ps> {
   Instruction Opcode = !cast<Instruction>(NAME);
   bit IsSingle = ps.Pfl.IsSingle;
+  bit IsInvalidSingleUseConsumer = ps.Pfl.IsInvalidSingleUseConsumer;
+  bit IsInvalidSingleUseProducer = ps.Pfl.IsInvalidSingleUseProducer;
```
Copy from the pseudo not the profile.

jayfoad commented on llvm/lib/Target/AMDGPU/VOPInstructions.td:
```diff
@@ -81,6 +83,8 @@ class VOP_Pseudo <string opName, string suffix, VOPProfile P, dag outs, dag ins,
   string Mnemonic = opName;
   Instruction Opcode = !cast<Instruction>(NAME);
   bit IsTrue16 = P.IsTrue16;
+  bit IsInvalidSingleUseConsumer = P.IsInvalidSingleUseConsumer;
+  bit IsInvalidSingleUseProducer = P.IsInvalidSingleUseProducer;
```
Just set them to 0 (or false) here and let specific instructions override them.

jayfoad reviewed (commented):
Just noticed some unnecessary stuff in this patch. Could you submit a follow up to clean it up please?
